tests/msc3083_test.go

```diff
@@ -181,6 +184,275 @@ func TestRestrictedRoomsRemoteJoin(t *testing.T) {
+	room := charlie.CreateRoom(t, map[string]interface{}{
+		"preset":       "public_chat",
+		"name":         "Room",
+		"room_version": "org.matrix.msc3083.v2",
+		"initial_state": []map[string]interface{}{
+			{
+				"type":      "m.room.join_rules",
+				"state_key": "",
+				"content": map[string]interface{}{
+					"join_rule": "restricted",
+					"allow": []map[string]interface{}{
+						{
+							"type":    "m.room_membership",
+							"room_id": &space,
+							"via":     []string{"hs2"},
+						},
+					},
+				},
+			},
+		},
+	})
+	charlie.SendEventSynced(t, space, b.Event{
+		Type:     "m.space.child",
+		StateKey: &room,
+		Content: map[string]interface{}{
+			"via": []string{"hs2"},
+		},
+	})
+
+	// Invite alice manually and accept it.
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	charlie.InviteRoom(t, room, alice.UserID)
+	alice.JoinRoom(t, room, []string{"hs2"})
+
+	// Alice should not be able to issue invites.
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+	body := map[string]interface{}{
+		"user_id": bob.UserID,
+	}
+	res := alice.DoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "r0", "rooms", room, "invite"},
+		client.WithJSONBody(t, body),
+	)
+	must.MatchResponse(t, res, match.HTTPResponse{
+		StatusCode: 403,
+	})
+
+	// Bob cannot join the room.
+	failJoinRoom(t, bob, room, "hs1", 403)
+
+	// Join the space, joining the room should work, but will bounce through hs2.
+	bob.JoinRoom(t, space, []string{"hs2"})
+	bob.JoinRoom(t, room, []string{"hs1"})
+
+	// Ensure that the join comes down sync on hs2. Note that we want to ensure hs2
+	// accepted the event.
+	charlie.SyncUntilTimelineHas(
+		t,
+		room,
+		func(ev gjson.Result) bool {
+			if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != bob.UserID {
+				return false
+			}
+			must.EqualStr(t, ev.Get("sender").Str, bob.UserID, "Bob should have joined by himself")
+			must.EqualStr(t, ev.Get("content").Get("membership").Str, "join", "Bob failed to join the room")
+
+			return true
+		},
+	)
+
+	// Raise the power level so that users on hs1 can invite people and then leave
+	// the room.
+	state_key := ""
+	charlie.SendEventSynced(t, room, b.Event{
+		Type:     "m.room.power_levels",
+		StateKey: &state_key,
+		Content: map[string]interface{}{
+			"invite": 0,
+			"users": map[string]interface{}{
+				charlie.UserID: 100,
+			},
+		},
+	})
+	charlie.LeaveRoom(t, room)
+
+	// Ensure the events have synced to hs1.
+	alice.SyncUntilTimelineHas(
+		t,
+		room,
+		func(ev gjson.Result) bool {
+			if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != charlie.UserID {
+				return false
+			}
+			must.EqualStr(t, ev.Get("content").Get("membership").Str, "leave", "Charlie failed to leave the room")
+
+			return true
+		},
+	)
+
+	// Have bob leave and rejoin. This should still work even though hs2 isn't in
+	// the room anymore!
+	bob.LeaveRoom(t, room)
```

Comment:
This should probably be synced with `alice.SyncUntilTimelineHas` as well right?

Reply:
Is there any reason this is just always synced? Seems like it would be an improvement.